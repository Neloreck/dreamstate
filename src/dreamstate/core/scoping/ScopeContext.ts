import { Context, createContext } from "react";

import { IRegistry } from "@/dreamstate/core/scoping/registry/createRegistry";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import {
  IBaseSignal,
  IContextManagerConstructor,
  IOptionalQueryRequest,
  ISignalEvent,
  TAnyContextManagerConstructor,
  TAnyObject,
  TAnyValue,
  TCallable,
  TQueryListener,
  TQueryResponse,
  TQueryType,
  TSignalListener,
  TUpdateObserver,
  TUpdateSubscriber,
} from "@/dreamstate/types";

/**
 * Internals of the scope context, meant for internal use only.
 *
 * This interface defines the internal properties and methods for managing data flow, signals,
 * and queries within the scope. It is not intended for external access and should only be used
 * by the `dreamstate` library for its internal operations.
 */
export interface IScopeContextInternals {
  /**
   * Mutable registry that contains all references and scope data.
   */
  REGISTRY: IRegistry;

  /**
   * Registers a `ManagerClass` in the current scope and creates an instance for context provision.
   *
   * @template T - Type of initial state injected into the manager.
   * @template C - Type of initial context injected into the manager.
   * @param {IContextManagerConstructor} ManagerClass - The manager class that should be registered in the scope.
   * @param {T | null} [initialState] - Optional initial state parameter that will be injected into the manager's
   *   constructor upon creation.
   * @param {Partial<C>} [initialContext] - Optional initial context parameter that will be injected into the
   *   manager and mixed with the context.
   * @returns {boolean} `true` if the service was successfully registered, `false` otherwise.
   */
  registerService<T extends TAnyObject, C extends TAnyObject>(
    ManagerClass: TAnyContextManagerConstructor,
    initialState?: T | null,
    initialContext?: Partial<C>
  ): boolean;

  /**
   * Unregisters a `ManagerClass` from the current scope.
   * Cleans up memory and removes all internal references associated with the manager.
   *
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class that should be unregistered from the scope.
   * @returns {boolean} `true` if the service was successfully unregistered, `false` otherwise.
   */
  unRegisterService(ManagerClass: TAnyContextManagerConstructor): boolean;

  /**
   * Adds an observer to a specified service.
   * The observer will be notified when the manager's context is updated.
   *
   * @template T - Type of the manager's context.
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class reference for which the observer
   *   should be added.
   * @param {TUpdateObserver} serviceObserver - The observer that should be notified on manager updates.
   * @param {number} [referencesCount] - Optional count of manager references for injection.
   *   By default, checks the registry.
   */
  addServiceObserver(
    ManagerClass: TAnyContextManagerConstructor,
    serviceObserver: TUpdateObserver,
    referencesCount?: number
  ): void;

  /**
   * Removes an observer from a specified service.
   * The observer will no longer be notified of updates from the manager's context.
   *
   * @template T - Type of the manager's context.
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class reference for which
   *   the observer should be removed.
   * @param {TUpdateObserver} serviceObserver - The observer that should be removed from the registry.
   * @param {number} [referencesCount] - Optional count of manager references for injection.
   *   By default, checks the registry.
   */
  removeServiceObserver(
    ManagerClass: TAnyContextManagerConstructor,
    serviceObserver: TUpdateObserver,
    referencesCount?: number
  ): void;

  /**
   * Notifies all observers that are subscribed to a manager's context updates.
   * This will update all providers based on the current context of the provided manager instance.
   *
   * @template T - Type of the manager's context.
   * @param {ContextManager<T>} manager - The manager instance whose context should trigger updates
   *   to subscribed providers.
   */
  notifyObservers<T extends TAnyObject>(manager: ContextManager<T>): void;

  /**
   * Subscribes to manager context updates.
   * The provided callback will be triggered whenever the manager's context is updated.
   *
   * @template T - Type of the manager's context.
   * @template D - Type of the context manager constructor.
   * @param {IContextManagerConstructor} ManagerClass - The manager class reference for which to subscribe.
   * @param {TUpdateSubscriber<T>} subscriber - The callback function that will be triggered on context updates.
   * @returns {TCallable} A function that can be used to unsubscribe from context updates.
   */
  subscribeToManager<T extends TAnyObject, D extends IContextManagerConstructor<TAnyValue, T>>(
    ManagerClass: D,
    subscriber: TUpdateSubscriber<T>
  ): TCallable;

  /**
   * Unsubscribes from manager context updates.
   *
   * @template T - Type of the manager's context.
   * @template D - Type of the context manager constructor.
   * @param {IContextManagerConstructor} ManagerClass - The manager class reference for which to unsubscribe.
   * @param {TUpdateSubscriber<T>} subscriber - The callback function that should be removed
   *   from context updates subscription.
   */
  unsubscribeFromManager<T extends TAnyObject, D extends IContextManagerConstructor<TAnyValue, T>>(
    ManagerClass: D,
    subscriber: TUpdateSubscriber<T>
  ): void;
}

/**
 * Represents the current scope context.
 * Provides public methods to interact with the current scope, including managing context, signals, and queries.
 */
export interface IScopeContext {
  /**
   * Library internals.
   * Not intended for normal use, except during unit testing.
   */
  INTERNAL: IScopeContextInternals;

  /**
   * Retrieves the current context snapshot for the given manager class.
   *
   * @template T - Type of the context.
   * @template D - Type of the context manager constructor.
   * @param manager - The manager class whose context is to be retrieved.
   * @returns The current context snapshot for the specified manager.
   */
  getContextOf<T extends TAnyObject, D extends IContextManagerConstructor<T>>(manager: D): T;

  /**
   * Emits a signal for all subscribers in the current Dreamstate scope.
   *
   * @template D - Type of the signal data.
   * @param base - The base signal object that triggers the event.
   *   It includes a `type` field  and an optional `data` field.
   * @param emitter - Optional signal emitter reference.
   * @returns A signal event instance that wraps the emitted signal.
   */
  emitSignal<D = undefined>(base: IBaseSignal<D>, emitter?: TAnyContextManagerConstructor | null): ISignalEvent<D>;

  /**
   * Subscribes to signals within the current scope.
   * The listener will be invoked on every signal event.
   *
   * @template D - Type of the signal data.
   * @param listener - The callback function to be invoked on signal events.
   * @returns A callable function that unsubscribes the listener.
   */
  subscribeToSignals<D = undefined>(listener: TSignalListener<D>): TCallable;

  /**
   * Unsubscribes a listener from signals in the current scope.
   * The provided listener will no longer be invoked on signal events.
   *
   * @template D - Type of the signal data.
   * @param listener - The callback function to remove from signal subscriptions.
   */
  unsubscribeFromSignals<D = undefined>(listener: TSignalListener<D>): void;

  /**
   * Registers a query provider callback to answer query data calls.
   *
   * @template T - Type of the query.
   * @param queryType - The type of query for data provisioning.
   * @param listener - The callback that listens to queries and returns evaluation data.
   * @returns A callable function that unregisters the query provider.
   */

  registerQueryProvider<T extends TQueryType>(queryType: T, listener: TQueryListener<T, TAnyValue>): TCallable;

  /**
   * Unregisters a query provider callback.
   * The callback will no longer handle query data calls.
   *
   * @template T - Type of the query.
   * @param queryType - The type of query for which the provider is unregistered.
   * @param listener - The callback to be unregistered.
   */
  unRegisterQueryProvider<T extends TQueryType>(queryType: T, listener: TQueryListener<T, TAnyValue>): void;

  /**
   * Queries data synchronously from the current scope.
   * The handler for the specified query type is executed, and its result is wrapped as a query response.
   *
   * @template D - Type of the data provided in the query request.
   * @template T - Type of the query.
   * @template Q - Type of the query request.
   * @param query - The query request containing the query type and optional data.
   * @returns The query response or null if no handler is found.
   */
  queryDataSync<D, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(query: Q): TQueryResponse<TAnyValue>;

  /**
   * Queries data asynchronously from the current scope.
   * The handler for the specified query type is executed, and its result is wrapped as a query response.
   *
   * @template D - Type of the data provided in the query request.
   * @template T - Type of the query.
   * @template Q - Type of the query request.
   * @param query - The query request containing the query type and optional data.
   * @returns A promise that resolves with the query response or null if no handler is found.
   */
  queryDataAsync<D, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(
    query: Q
  ): Promise<TQueryResponse<TAnyValue>>;
}

/**
 * React context reference for the Dreamstate scope, used to wrap the virtual DOM tree.
 * This context allows the React tree to be aware of the current scope and provides access
 * to context management functionalities within the Dreamstate data flow.
 */
export const ScopeContext: Context<IScopeContext> = createContext(null as unknown as IScopeContext);
