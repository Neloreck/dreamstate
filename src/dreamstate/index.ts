/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export {
  ILoadable as Loadable,
  TNested as Nested,
  TComputed as Computed,
  TStateSetter as StateSetter,
  TConsumable as Consumable,
  ISignalEvent as SignalEvent,
  ISignal as Signal,
  TSignalType as SignalType,
  TQueryResponse as QueryResponse,
  TOptionalQueryResponse as OptionalQueryResponse,
  TQueryRequest as QueryRequest,
  TQueryType as QueryType
} from "@/dreamstate/types";

export { ContextService } from "@/dreamstate/core/services/ContextService";
export { ContextManager } from "@/dreamstate/core/services/ContextManager";

export { getCurrentContext } from "@/dreamstate/core/registry/getCurrentContext";
export { getCurrent } from "@/dreamstate/core/registry/getCurrent";

export { createProvider } from "@/dreamstate/core/provision/createProvider";
export { Provide } from "@/dreamstate/core/provision/Provide";
export { withProvision } from "@/dreamstate/core/provision/withProvision";

export { OnSignal } from "@/dreamstate/core/signals/OnSignal";
export { useSignals } from "@/dreamstate/core/signals/useSignals";
export { unsubscribeFromSignals } from "@/dreamstate/core/signals/unsubscribeFromSignals";
export { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
export { emitSignal } from "@/dreamstate/core/signals/emitSignal";

export { OnQuery } from "@/dreamstate/core/queries/OnQuery";
export { queryData } from "@/dreamstate/core/queries/queryData";

export { useManager } from "@/dreamstate/core/consumption/useManager";
export { Consume } from "@/dreamstate/core/consumption/Consume";
export { withConsumption } from "@/dreamstate/core/consumption/withConsumption";

export { Bind } from "@/dreamstate/utils/Bind";
export { createLoadable } from "@/dreamstate/utils/createLoadable";
export { createNested } from "@/dreamstate/utils/createNested";
export { createComputed } from "@/dreamstate/utils/createComputed";
export { createSetter } from "@/dreamstate/utils/createSetter";
