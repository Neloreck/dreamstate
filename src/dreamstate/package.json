{
  "name": "dreamstate",
  "version": "4.0.0-experimental.6",
  "description": "Store management library based on context and observers.",
  "sideEffects": false,
  "author": "Syrotenko Igor",
  "homepage": "https://github.com/neloreck/dreamstate#readme",
  "license": "MIT",
  "main": "./index.js",
  "module": "./esm/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/neloreck/dreamstate.git"
  },
  "bugs": {
    "url": "https://github.com/neloreck/dreamstate/issues"
  },
  "keywords": [
    "action",
    "class-based",
    "declarative",
    "decorator",
    "metadata",
    "react",
    "reducer",
    "redux",
    "state",
    "typescript",
    "utility"
  ],
  "dependencies": {
    "tslib": "^2.1.0",
    "shallow-equal": ">=1.0.0"
  },
  "peerDependencies": {
    "react": ">=16.8.0"
  }
}
