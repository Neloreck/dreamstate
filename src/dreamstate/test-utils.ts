/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export { ScopeContext as InternalScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

export * from "@/dreamstate/test-utils/registry/getCurrent";
export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";

export * from "@/dreamstate/test-utils/registry/mockManagerWithScope";
export * from "@/dreamstate/test-utils/registry/mockScopeProvider";
export * from "@/dreamstate/test-utils/registry/mockScope";

export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";
