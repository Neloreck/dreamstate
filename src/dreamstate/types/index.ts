export * from "@/dreamstate/types/queries";
export * from "@/dreamstate/types/signals";
export * from "@/dreamstate/types/error";
export * from "@/dreamstate/types/general";
export * from "@/dreamstate/types/decorators";
export * from "@/dreamstate/types/utils";
export * from "@/dreamstate/types/internal";
