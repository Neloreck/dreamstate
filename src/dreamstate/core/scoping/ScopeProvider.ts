import { createElement, ProviderProps, ReactElement, ReactNode, useState } from "react";

import { initializeScopeContext } from "@/dreamstate/core/scoping/initializeScopeContext";
import { IScopeContext, ScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

/**
 * Lazy initializer of current scope context provider props object that preserves object references.
 * Composes props object that will be same for each tree re-render.
 */
function scopeStateInitializer(): ProviderProps<IScopeContext> {
  return { value: initializeScopeContext() };
}

export interface IScopeProviderProps {
  children?: ReactNode;
}

/**
 * Dreamstate scope provider.
 * Ensures that signaling and context managers are isolated under this specific provider.
 *
 * @param {IScopeProviderProps} props - react component properties that contain provider children.
 *
 * @return {ReactElement} react tree element.
 */
export function ScopeProvider(props: IScopeProviderProps): ReactElement {
  const [ scopeContextProviderProps ] = useState(scopeStateInitializer);

  return createElement(ScopeContext.Provider, scopeContextProviderProps, props.children);
}

/**
 * Easier devtools usage for dev environment.
 */
if (IS_DEV) {
  ScopeProvider.displayName = "Dreamstate.ScopeProvider";
}
