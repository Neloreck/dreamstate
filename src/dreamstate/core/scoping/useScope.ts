import { useContext } from "react";

import { IScopeContext, ScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

/**
 * Use current scope for current signaling/queries methods access.
 * Return bundle of functions that work with current scope and allow signals or queries processing.
 *
 * @return {IScopeContext} current scope in react tree.
 */
export function useScope(): IScopeContext {
  /**
   * Wrapper that encapsulates useContext + ScopeContext logic.
   */
  return useContext(ScopeContext);
}
