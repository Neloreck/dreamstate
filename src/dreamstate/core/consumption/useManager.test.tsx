import { mount } from "enzyme";
import { default as React, ReactElement } from "react";
import { act } from "react-dom/test-utils";

import { ContextManager, createProvider, ScopeProvider, useManager, useScope } from "@/dreamstate";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TUninitializedValue } from "@/dreamstate/types";

describe("useManager subscription and rendering", () => {
  const onStarted = jest.fn();
  const onEnded = jest.fn();

  interface ISampleContext {
    example: number;
    text: string;
  }

  class SampleContextManager extends ContextManager<ISampleContext> {
    public context: ISampleContext = {
      example: 1000,
      text: "value",
    };

    public onProvisionStarted(): void {
      onStarted();
    }

    public onProvisionEnded(): void {
      onEnded();
    }
  }

  const SampleProvider = createProvider([SampleContextManager]);

  afterEach(() => {
    onEnded.mockClear();
    onStarted.mockClear();
  });

  it("should properly fire provision events for functional subscribers and clean-up memory", async () => {
    let rendersCount: number = 0;
    let stateScope: IScopeContext = null as TUninitializedValue;

    function SampleConsumer(): ReactElement {
      const scope: IScopeContext = useScope();
      const value: ISampleContext = useManager(SampleContextManager);

      rendersCount += 1;
      stateScope = scope;

      return <span>{JSON.stringify(value)}</span>;
    }

    const tree = mount(
      <ScopeProvider>
        <SampleProvider>
          <SampleConsumer/>
        </SampleProvider>
      </ScopeProvider>
    );

    expect(tree.render()).toMatchSnapshot();

    expect(onStarted).toHaveBeenCalledTimes(1);
    expect(onEnded).toHaveBeenCalledTimes(0);
    expect(stateScope.INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY.get(SampleContextManager)).toBeDefined();
    expect(stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)).toBeDefined();

    act(() => {
      stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)!.setContext({ example: -1 });
    });
    tree.update();

    expect(tree).toMatchSnapshot();
    expect(rendersCount).toBe(2);

    act(() => {
      stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)!.setContext({
        text: "anything",
      });
    });
    tree.update();

    expect(tree).toMatchSnapshot();
    expect(rendersCount).toBe(3);

    tree.unmount();

    expect(onStarted).toHaveBeenCalledTimes(1);
    expect(onEnded).toHaveBeenCalledTimes(1);
    expect(rendersCount).toBe(3);

    expect(stateScope.INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY.get(SampleContextManager)).toBeUndefined();
    expect(stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)).toBeUndefined();
  });

  it("should properly handle memoized subscriptions", async () => {
    let rendersCount: number = 0;
    let stateScope: IScopeContext = null as TUninitializedValue;

    function SampleConsumer(): ReactElement {
      const scope: IScopeContext = useScope();
      const value: ISampleContext = useManager(SampleContextManager, ({ text }) => [text]);

      rendersCount += 1;
      stateScope = scope;

      return <span>{JSON.stringify(value)}</span>;
    }

    const tree = mount(
      <ScopeProvider>
        <SampleProvider>
          <SampleConsumer/>
        </SampleProvider>
      </ScopeProvider>
    );

    expect(tree.render()).toMatchSnapshot();

    act(() => {
      stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)!.setContext({ example: -1 });
    });
    tree.update();

    expect(tree).toMatchSnapshot();
    expect(rendersCount).toBe(1);

    act(() => {
      stateScope.INTERNAL.REGISTRY.CONTEXT_INSTANCES_REGISTRY.get(SampleContextManager)!.setContext({
        text: "anything",
      });
    });
    tree.update();

    expect(tree).toMatchSnapshot();
    expect(rendersCount).toBe(2);

    tree.unmount();

    expect(rendersCount).toBe(2);
  });
});
