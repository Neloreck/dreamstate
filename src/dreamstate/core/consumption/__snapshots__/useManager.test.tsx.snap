// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (combined) 1`] = `
<span>
  {"example":1000,"text":"value"}
</span>
`;

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (combined) 2`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.CombinedProvider[SampleContextManager]>
    <SampleConsumer>
      <span>
        {"example":-1,"text":"value"}
      </span>
    </SampleConsumer>
  </Dreamstate.CombinedProvider[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (combined) 3`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.CombinedProvider[SampleContextManager]>
    <SampleConsumer>
      <span>
        {"example":-1,"text":"anything"}
      </span>
    </SampleConsumer>
  </Dreamstate.CombinedProvider[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (scoped) 1`] = `
<span>
  {"example":1000,"text":"value"}
</span>
`;

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (scoped) 2`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.ScopedProviders[SampleContextManager]>
    <Dreamstate.ScopedProvider
      ManagerClass={[Function]}
      dependencies={
        Array [
          [Function],
        ]
      }
      scope={
        Object {
          "INTERNAL": Object {
            "REGISTRY": Object {
              "CONTEXT_INSTANCES_REGISTRY": Map {
                [Function] => SampleContextManager {
                  "IS_DISPOSED": false,
                  "context": Object {
                    "example": -1,
                    "text": "value",
                  },
                  Symbol(SCOPE): [Circular],
                  Symbol(SIGNALING_HANDLER): [Function],
                  Symbol(SIGNAL_METADATA): Array [],
                  Symbol(QUERY_METADATA): Array [],
                },
              },
              "CONTEXT_OBSERVERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "CONTEXT_SERVICES_REFERENCES": Map {
                [Function] => 1,
              },
              "CONTEXT_STATES_REGISTRY": Map {
                [Function] => Object {
                  "example": -1,
                  "text": "value",
                },
              },
              "CONTEXT_SUBSCRIBERS_REGISTRY": Map {
                [Function] => Set {},
              },
              "QUERY_PROVIDERS_REGISTRY": Map {},
              "SIGNAL_LISTENERS_REGISTRY": Set {
                [Function],
              },
            },
            "addServiceObserver": [Function],
            "notifyObservers": [Function],
            "registerManager": [Function],
            "removeServiceObserver": [Function],
            "subscribeToManager": [Function],
            "unRegisterManager": [Function],
            "unsubscribeFromManager": [Function],
          },
          "emitSignal": [Function],
          "getContextOf": [Function],
          "getInstanceOf": [Function],
          "queryDataAsync": [Function],
          "queryDataSync": [Function],
          "registerQueryProvider": [Function],
          "subscribeToSignals": [Function],
          "unRegisterQueryProvider": [Function],
          "unsubscribeFromSignals": [Function],
        }
      }
    >
      <SampleConsumer>
        <span>
          {"example":-1,"text":"value"}
        </span>
      </SampleConsumer>
    </Dreamstate.ScopedProvider>
  </Dreamstate.ScopedProviders[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly fire provision events for functional subscribers and clean-up memory (scoped) 3`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.ScopedProviders[SampleContextManager]>
    <Dreamstate.ScopedProvider
      ManagerClass={[Function]}
      dependencies={
        Array [
          [Function],
        ]
      }
      scope={
        Object {
          "INTERNAL": Object {
            "REGISTRY": Object {
              "CONTEXT_INSTANCES_REGISTRY": Map {
                [Function] => SampleContextManager {
                  "IS_DISPOSED": false,
                  "context": Object {
                    "example": -1,
                    "text": "anything",
                  },
                  Symbol(SCOPE): [Circular],
                  Symbol(SIGNALING_HANDLER): [Function],
                  Symbol(SIGNAL_METADATA): Array [],
                  Symbol(QUERY_METADATA): Array [],
                },
              },
              "CONTEXT_OBSERVERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "CONTEXT_SERVICES_REFERENCES": Map {
                [Function] => 1,
              },
              "CONTEXT_STATES_REGISTRY": Map {
                [Function] => Object {
                  "example": -1,
                  "text": "anything",
                },
              },
              "CONTEXT_SUBSCRIBERS_REGISTRY": Map {
                [Function] => Set {},
              },
              "QUERY_PROVIDERS_REGISTRY": Map {},
              "SIGNAL_LISTENERS_REGISTRY": Set {
                [Function],
              },
            },
            "addServiceObserver": [Function],
            "notifyObservers": [Function],
            "registerManager": [Function],
            "removeServiceObserver": [Function],
            "subscribeToManager": [Function],
            "unRegisterManager": [Function],
            "unsubscribeFromManager": [Function],
          },
          "emitSignal": [Function],
          "getContextOf": [Function],
          "getInstanceOf": [Function],
          "queryDataAsync": [Function],
          "queryDataSync": [Function],
          "registerQueryProvider": [Function],
          "subscribeToSignals": [Function],
          "unRegisterQueryProvider": [Function],
          "unsubscribeFromSignals": [Function],
        }
      }
    >
      <SampleConsumer>
        <span>
          {"example":-1,"text":"anything"}
        </span>
      </SampleConsumer>
    </Dreamstate.ScopedProvider>
  </Dreamstate.ScopedProviders[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (combined) 1`] = `
<span>
  {"example":1000,"text":"value"}
</span>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (combined) 2`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.CombinedProvider[SampleContextManager]>
    <SampleConsumer>
      <span>
        {"example":1000,"text":"value"}
      </span>
    </SampleConsumer>
  </Dreamstate.CombinedProvider[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (combined) 3`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.CombinedProvider[SampleContextManager]>
    <SampleConsumer>
      <span>
        {"example":-1,"text":"anything"}
      </span>
    </SampleConsumer>
  </Dreamstate.CombinedProvider[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (scoped) 1`] = `
<span>
  {"example":1000,"text":"value"}
</span>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (scoped) 2`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.ScopedProviders[SampleContextManager]>
    <Dreamstate.ScopedProvider
      ManagerClass={[Function]}
      dependencies={
        Array [
          [Function],
        ]
      }
      scope={
        Object {
          "INTERNAL": Object {
            "REGISTRY": Object {
              "CONTEXT_INSTANCES_REGISTRY": Map {
                [Function] => SampleContextManager {
                  "IS_DISPOSED": false,
                  "context": Object {
                    "example": -1,
                    "text": "value",
                  },
                  Symbol(SCOPE): [Circular],
                  Symbol(SIGNALING_HANDLER): [Function],
                  Symbol(SIGNAL_METADATA): Array [],
                  Symbol(QUERY_METADATA): Array [],
                },
              },
              "CONTEXT_OBSERVERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "CONTEXT_SERVICES_REFERENCES": Map {
                [Function] => 1,
              },
              "CONTEXT_STATES_REGISTRY": Map {
                [Function] => Object {
                  "example": -1,
                  "text": "value",
                },
              },
              "CONTEXT_SUBSCRIBERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "QUERY_PROVIDERS_REGISTRY": Map {},
              "SIGNAL_LISTENERS_REGISTRY": Set {
                [Function],
              },
            },
            "addServiceObserver": [Function],
            "notifyObservers": [Function],
            "registerManager": [Function],
            "removeServiceObserver": [Function],
            "subscribeToManager": [Function],
            "unRegisterManager": [Function],
            "unsubscribeFromManager": [Function],
          },
          "emitSignal": [Function],
          "getContextOf": [Function],
          "getInstanceOf": [Function],
          "queryDataAsync": [Function],
          "queryDataSync": [Function],
          "registerQueryProvider": [Function],
          "subscribeToSignals": [Function],
          "unRegisterQueryProvider": [Function],
          "unsubscribeFromSignals": [Function],
        }
      }
    >
      <SampleConsumer>
        <span>
          {"example":1000,"text":"value"}
        </span>
      </SampleConsumer>
    </Dreamstate.ScopedProvider>
  </Dreamstate.ScopedProviders[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;

exports[`useManager subscription and rendering should properly handle memoized subscriptions (scoped) 3`] = `
<Dreamstate.ScopeProvider>
  <Dreamstate.ScopedProviders[SampleContextManager]>
    <Dreamstate.ScopedProvider
      ManagerClass={[Function]}
      dependencies={
        Array [
          [Function],
        ]
      }
      scope={
        Object {
          "INTERNAL": Object {
            "REGISTRY": Object {
              "CONTEXT_INSTANCES_REGISTRY": Map {
                [Function] => SampleContextManager {
                  "IS_DISPOSED": false,
                  "context": Object {
                    "example": -1,
                    "text": "anything",
                  },
                  Symbol(SCOPE): [Circular],
                  Symbol(SIGNALING_HANDLER): [Function],
                  Symbol(SIGNAL_METADATA): Array [],
                  Symbol(QUERY_METADATA): Array [],
                },
              },
              "CONTEXT_OBSERVERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "CONTEXT_SERVICES_REFERENCES": Map {
                [Function] => 1,
              },
              "CONTEXT_STATES_REGISTRY": Map {
                [Function] => Object {
                  "example": -1,
                  "text": "anything",
                },
              },
              "CONTEXT_SUBSCRIBERS_REGISTRY": Map {
                [Function] => Set {
                  [Function],
                },
              },
              "QUERY_PROVIDERS_REGISTRY": Map {},
              "SIGNAL_LISTENERS_REGISTRY": Set {
                [Function],
              },
            },
            "addServiceObserver": [Function],
            "notifyObservers": [Function],
            "registerManager": [Function],
            "removeServiceObserver": [Function],
            "subscribeToManager": [Function],
            "unRegisterManager": [Function],
            "unsubscribeFromManager": [Function],
          },
          "emitSignal": [Function],
          "getContextOf": [Function],
          "getInstanceOf": [Function],
          "queryDataAsync": [Function],
          "queryDataSync": [Function],
          "registerQueryProvider": [Function],
          "subscribeToSignals": [Function],
          "unRegisterQueryProvider": [Function],
          "unsubscribeFromSignals": [Function],
        }
      }
    >
      <SampleConsumer>
        <span>
          {"example":-1,"text":"anything"}
        </span>
      </SampleConsumer>
    </Dreamstate.ScopedProvider>
  </Dreamstate.ScopedProviders[SampleContextManager]>
</Dreamstate.ScopeProvider>
`;
